### DISCUSSION TITLE
Abstraction Patterns for useQuery states

### DISCUSSION THREAD

**Discussion by rohitpotato** [TanStack/query#1826] [Q&A] [answered]
The most typical way of rendering the component goes something like this:

```
const query = useQuery(...)

if(query.isLoading) {
  return 'Loading...'
}

if(query.isError) {
  return 'Error...'
}

return <div>{REAL_QUERY_DATA}</div>
```

This logic has to repeated for every component. 

Is there a way to extract this logic? What I mean by that is showing a loader until the query data loads and then displaying the real component with the data. 

Any such patterns where we dont have to write multiple `if statements` and directly render the children while maintaining the ```loading``` and error ```states```.

Thanks in advance!

**Comment by TkDodo** [comment, marked as answer]
Enabling `suspense` would show the `fallback` for loading and uses `ErrorBoundary` for errors.

**Reply by rohitpotato** [reply]
> Enabling `suspense` would show the `fallback` for loading and uses `ErrorBoundary` for errors.

True but `suspense` is still experimental. Is there any other way around to handle such use cases? Something like for each component using `useQuery` we can wrap that component in a container that would render` loading, error and sucess states`?

What approaches are preferable for production?

**Reply by TkDodo** [reply]
ErrorBoundaries can be used w/o suspense. You can also write a HOC but I’ve had different use-cases for loading states. Some have initialData, not all show the same spinner etc..

**Reply by rohitpotato** [reply]
How would you go about writing a reusable HOC? Any suggestions? Because my use-case is pretty generic. Same loading states everywhere and almost same configuration for every query.

**Reply by TkDodo** [reply]
render props:

```
const ReactQueryWrapper = ({ children, ...props }) => {
  const { isLoading, data, isError, error } = useQuery(props)
  if (isLoading) return 'Loading...'
  if (isError) return 'Error: ' + error

  return children(data)
}
```

```
<ReactQueryWrapper queryFn={queryFn} queryKey="todos" >
    {data => ...}
</ReactQueryWrapper>

**Reply by rohitpotato** [reply]
Just a quick follow up question, why wont something like this work? I guess this is more of a react question than react-query.

```
const Wrapper = ({ query, LoadingComponent, ErrorComponent, children }) => {
  if (query.isLoading) {
    const toRender = LoadingComponent || <DefaultLoader />;
    return <div className="grid place-items-center">{toRender}</div>;
  }

  if (query.isError) {
    const toRender = ErrorComponent ? (
      <ErrorComponent />
    ) : (
      <div className="dark:text-white">Failed to Load</div>
    );
    return <div className="grid place-items-center">{toRender}</div>;
  }

  if (query.isSuccess) {
    return React.Children.only(children);
  }
  return null;
};
```
And using it like:

```
<Wrapper query={query}>
   <div>{query.data.message}</div>
</Wrapper>
```

**Reply by TkDodo** [reply]
also works, but the children of `wrapper` are executed every time, so you have to work with possibly undefined data. At least, it must read:

```
<Wrapper query={query}>
   <div>{query.data?.message}</div>
</Wrapper>
```

**Reply by rohitpotato** [reply]
Ya that explains why it was working for some components and not for others, although I did understand the children are executed everytime. Does that mean its better to stick with `renderProps` for performance benifits?

**Reply by TkDodo** [reply]
not for perf, but maybe for convenience.

**Reply by rohitpotato** [reply]
I have noticed that even though the children are executed everytime, they are not visible on the UI. Why is that?

Check out this sandbox for more clarity y: https://codesandbox.io/s/react-composition-and-render-prop-tyyzh?file=/src/App.js

And Thank you @TkDodo for keeping up with this conversation. This surely helped in clearing some of the fundamentals for me.

**Reply by TkDodo** [reply]
> I have noticed that even though the children are executed everytime, they are not visible on the UI. Why is that?

because you only return the children from the wrapper component inside of `if (query.isSuccess)`. components are just functions. The children are created and passed to the wrapper, but if the wrapper chooses to discard them, then they are not visible

**Reply by rohitpotato** [reply]
Hmmm, it seems like I am not able to grasp the concept properly. I am thinking if the children are **executed** at every render resulting in `Cant read property of undefined` , then why does the the `Parent` component does not render those children even after I do `optional chaining`.  In the sandbox we can see the logs, that means at every render it's executing. 

And, if the `Wrapper` **only** renders the children `if(query.isSuccess)` then the data should be available.

**My confusion is that if the logs are appearing, then why not the actual elements that are also being executed.**

```
const Parent = () => {
    const someData  = useQuery();
    return (
        <Wrapper {...someData}>
              <div>{someData?.message}</div>
       <Wrapper />
    )
}
```

**Reply by TkDodo** [reply]
> My confusion is that if the logs are appearing, then why not the actual elements that are also being executed.

They are "executed". Let me give you a different example:

```
function MyComponent() {
  const compute = () => 'Hello World'
   
  const result = compute()
  return 'Foo Bar'
}
```

If you look at this code, you will see that the function named `compute` is executed - if you put a log there, it will print, but if you then take a look at the result of `MyComponent`, it is nowhere to be found, because I _discard_ it with the return statement. Basically, I am unnecessarily running it.

Now a JSX expression is nothing more than a function call:

```
<div>Hello World
```

is the same as

```
React.createElement('div', {}, 'Hello World')
```

the function call will be executed, but what happens with the result of it depends on what you do with `children`. The bottom line is: if you have a component that doesn't _return_ `children`, they will not be shown.

**Reply by rohitpotato** [reply]
@TkDodo really good explanation. It didn't hit me and I completely forgot about `React.createElement`. Thanks for having this long conversation. Really did help me out and hopefully will help others to come up with more abstraction patterns.

The render prop method looks good to me and my requirement, I'll be using that. :)
